### PR TITLE
Allow the uncompress depacker to work without using a temporary file

### PR DESCRIPTION
**Pull request by ccawley2011**

### CODE REVIEW

**Comment by sezero** [comment]
What the hell is wrong with CI cmake MSVC builds?
```
Run cmd.exe /C "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64 "&&" cmake . -DCMAKE_BUILD_TYPE=Debug -DWITH_UNIT_TESTS=ON -B build
  cmd.exe /C "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64 "&&" cmake . -DCMAKE_BUILD_TYPE=Debug -DWITH_UNIT_TESTS=ON -B build
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
Error: Process completed with exit code 1.
```

**Comment by ccawley2011** [comment]
> What the hell is wrong with CI cmake MSVC builds?
> 
> ```
> Run cmd.exe /C "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64 "&&" cmake . -DCMAKE_BUILD_TYPE=Debug -DWITH_UNIT_TESTS=ON -B build
>   cmd.exe /C "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64 "&&" cmake . -DCMAKE_BUILD_TYPE=Debug -DWITH_UNIT_TESTS=ON -B build
>   shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
> 'C:\Program' is not recognized as an internal or external command,
> operable program or batch file.
> Error: Process completed with exit code 1.
> ```

I'm not 100% sure, but it seems to be related to the PR being built using the `windows-2022` build environment instead of `windows-2019`.

File: src/depackers/uncompress.c
Comment (AliceLR) [inline src/depackers/uncompress.c:248]: Pretty much all of the streaming depackers need a slightly different approach here to avoid slow loads and the Linux OOM killer. Instead of expanding the output buffer linearly, it should be expanded exponentially and bounded against `LIBXMP_DEPACK_LIMIT`. Something like this *might* work (there's a `next_power_of_two` in bunzip2.c that can be copied):

```C
...
if (outpos >= outsize) {
  outsize = next_power_of_two(outpos);
  if (outsize > LIBXMP_DEPACK_LIMIT) {
    free(outbuf);
    return -1;
  }
  tmp = (char_type *) realloc(outbuf, outsize + 2048);
  ...
```

Comment (AliceLR) [reply src/depackers/uncompress.c:248]: I'll check my fuzzing corpus for any existing files that generate slow loads with this patch just in case.

**Comment by AliceLR** [comment]
> What the hell is wrong with CI cmake MSVC builds?

> I'm not 100% sure, but it seems to be related to the PR being built using the `windows-2022` build environment instead of `windows-2019`.

To add to this: I was getting these on my own branches before, but not on PRs. But it looks like it's affecting the PR regression tests now as well? Fantastic...

File: src/depackers/uncompress.c
Comment (AliceLR) [reply src/depackers/uncompress.c:248]: Nope, nothing came up. I think this can just be left as-is for now, I want to see if libFuzzer turns up anything like it did for the others.

**Review by AliceLR — APPROVED** [review]
This looks OK to me. Any slow load tweaks can be done after if libFuzzer manages to find something.

**Comment by sezero** [comment]
Should we wait until the CI fix is updated and merged, so this can be rebased on top of it?

**Comment by AliceLR** [comment]
I think that would be OK, yeah.

**Comment by sezero** [comment]
OK, waiting for @ccawley2011 to fix the CI patch rebasing of this one against it after its merge.

**Comment by sezero** [comment]
OK, this is in.

**Pull request merged**
